syncstorage/src/tokenserver/handlers.rs

```diff
@@ -108,28 +112,72 @@ fn get_token_plaintext(
-    let keys_changed_at =
-        cmp::max(req.auth_data.keys_changed_at, req.user.keys_changed_at).unwrap_or(0);
-
-    let generation = if let Some(generation) = req.auth_data.generation {
-        // If there's a generation on the request, choose the larger of that and the generation
-        // already stored on the user record.
-        cmp::max(generation, req.user.generation)
-    } else {
-        // If there's not a generation on the request and the keys_changed_at on the request is
-        // larger than the generation stored on the user record, set the user's generation to be
-        // the keys_changed_at on the request.
-        cmp::max(req.auth_data.keys_changed_at, Some(req.user.generation)).unwrap_or(0)
+    let keys_changed_at = match (req.auth_data.keys_changed_at, req.user.keys_changed_at) {
+        // If the keys_changed_at in the request is larger than that stored on the user record,
+        // update to the value in the request.
+        (Some(request_keys_changed_at), Some(user_keys_changed_at))
+            if request_keys_changed_at >= user_keys_changed_at =>
+        {
+            Some(request_keys_changed_at)
+        }
+        // If there is a keys_changed_at in the request and it's smaller than that stored on the
+        // user record, we've already returned an error at this point.
+        (Some(_request_keys_changed_at), Some(_user_keys_changed_at)) => unreachable!(),
+        // If there is a keys_changed_at on the request but not one on the user record, this is the
+        // first time the client reported it, so we assign the new value.
+        (Some(request_keys_changed_at), None) => Some(request_keys_changed_at),
+        // At this point, we've already validated that, if there is a keys_changed_at already
+        // stored on the user record, there must be one in the request. If that isn't the case,
+        // we've already returned an error.
+        (None, Some(user_keys_changed_at)) if user_keys_changed_at != 0 => unreachable!(),
+        // If there's no keys_changed_at in the request and the keys_changed_at on the user record
+        // is 0, keep the value as 0.
+        (None, Some(_user_keys_changed_at)) => Some(0),
+        // If there is no keys_changed_at on the user record or in the request, we want to leave
+        // the value unset.
+        (None, None) => None,
+    };
+
+    let generation = match req.auth_data.generation {
+        // If there's a generation in the request and it's greater than or equal to that stored on
+        // the user record, update to the value in the request.
+        Some(request_generation) if request_generation >= req.user.generation => request_generation,
+        // If there's a generation in the request and it's smaller than that stored on the user
+        // record, we've already returned an error.
+        Some(_request_generation) => unreachable!(),
+        None => match (req.auth_data.keys_changed_at, req.user.keys_changed_at) {
```

Comment:
I thought very hard about the best way to organize this section to make it as readable as possible, but I am very open to suggestions if there's a clearer way!

Reply:
I like the idea of the heavily commented `match` case, mostly because it forces all the cases to be addressed.